cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DatastoreImpl.java

```diff
@@ -1525,95 +1525,81 @@ boolean isOpen() {
      *
      * @see <a href="http://wiki.apache.org/couchdb/HttpPostRevsDiff">HttpPostRevsDiff documentation</a>
      * @param revisions a Multimap of document id → revision id
-     * @return a Map of document id → collection of revision id: the subset of given the document
-     * id/revisions that are not stored in the database
+     * @return the subset of given the document id/revisions that are already stored in the database
      */
-    public Map<String, Collection<String>> revsDiff(final Multimap<String, String> revisions) {
+    public Map<String, List<String>> revsDiff(final Map<String, List<String>> revisions) {
         Preconditions.checkState(this.isOpen(), "Database is closed");
         Preconditions.checkNotNull(revisions, "Input revisions must not be null");
 
         try {
-            return queue.submit(new SQLCallable<Map<String,Collection<String>>>(){
+            return queue.submit(new SQLCallable<Map<String, List<String>>>() {
                 @Override
-                public Map<String, Collection<String>> call(SQLDatabase db) throws Exception {
-                    Multimap<String, String> missingRevs = ArrayListMultimap.create();
-                    // Break the potentially big multimap into small ones so for each map,
-                    // a single query can be use to check if the <id, revision> pairs in sqlDb or not
-                    List<Multimap<String, String>> batches =
-                            multiMapPartitions(revisions, SQLITE_QUERY_PLACEHOLDERS_LIMIT);
-                    for(Multimap<String, String> batch : batches) {
-                        revsDiffBatch(db,batch);
-                        missingRevs.putAll(batch);
+                public Map<String, List<String>> call(SQLDatabase db) throws Exception {
+
+                    ValueListMap<String, String> missingRevs = new ValueListMap<String, String>();
+
+                    // Break down by docId first to avoid potential rev ID clashes between doc IDs
+                    for (Map.Entry<String, List<String>> entry : revisions.entrySet()) {
+                        String docId = entry.getKey();
+                        List<String> revs = entry.getValue();
+                        // Partition into batches to avoid exceeding placeholder limit
+                        List<List<String>> batches = CollectionUtils.partition(revs,
+                                SQLITE_QUERY_PLACEHOLDERS_LIMIT - 1);
```

Comment:
I think we need to make a comment saying why its `SQLITE_QUERY_PLACEHOLDERS_LIMIT - 1` rather than just `SQLITE_QUERY_PLACEHOLDERS_LIMIT`